### PR TITLE
feat: Implement caching

### PR DESCRIPTION
**Pull request by rebasedming**
# Ticket(s) Closed

- Closes #

## What
`CREATE FOREIGN TABLE` now has a `cache` option. If set to true, a DuckDB table instead of view is created. This stores the table in DuckDB format inside the Postgres data directory.

```sql
CREATE FOREIGN DATA WRAPPER parquet_wrapper
HANDLER parquet_fdw_handler VALIDATOR parquet_fdw_validator;

-- Provide S3 credentials
CREATE SERVER parquet_server FOREIGN DATA WRAPPER parquet_wrapper;

-- Create foreign table with auto schema creation
CREATE FOREIGN TABLE trips ()
SERVER parquet_server
OPTIONS (files 's3://paradedb-benchmarks/yellow_tripdata_2024-01.parquet', cache 'true');

-- Success! Now you can query the remote Parquet file like a regular Postgres table
SELECT COUNT(*) FROM trips;
```

## Why
Enable caching for faster queries, sets us up for upserts later on.

## How
Every Postgres database now creates a DuckDB database. Connections are opened against this database.

### CODE REVIEW

**Comment by neilyio** [comment]
Now that the DuckDB database is on disk, how does this work with multiple Postgres connection processes accessing it simultaneously? My impression is that DuckDB needs to be in read-only mode to support multiple connections:

https://github.com/duckdb/duckdb/issues/1343
https://github.com/duckdb/duckdb/issues/40

If you feel like it's appropriate to add a test for this, you can make multiple instances of PgConnection in the test suite and try running queries with both.

**Comment by rebasedming** [comment]
> Now that the DuckDB database is on disk, how does this work with multiple Postgres connection processes accessing it simultaneously? My impression is that DuckDB needs to be in read-only mode to support multiple connections:
> 
> [duckdb/duckdb#1343](https://github.com/duckdb/duckdb/issues/1343) [duckdb/duckdb#40](https://github.com/duckdb/duckdb/issues/40)
> 
> If you feel like it's appropriate to add a test for this, you can make multiple instances of PgConnection in the test suite and try running queries with both.

Great catch. I had missed this caveat but I now see it https://duckdb.org/docs/connect/concurrency.

Do you think a background worker would solve?

**Review by philippemnoel — COMMENTED** [review]
Does this also cache the Parquet metadata?

File: @ [hunk 1]
```diff
@@ -0,0 +1,820 @@
+# -----------------------------
```
Comment (philippemnoel) [inline @:1]: Is this file a mistake?

File: src/lib.rs [hunk 2]
```diff
@@ -43,6 +44,7 @@ pub extern "C" fn _PG_init() {
     };
 
     GUCS.init("pg_analytics");
+    pg_shmem_init!(env::DUCKDB_CONNECTION_CACHE);
 
     // TODO: Change to ParadeExtension::PgAnalytics
```
Comment (philippemnoel) [inline src/lib.rs:49]: I've just noticed this here. I've updated this in my PR in the monorepo: https://github.com/paradedb/paradedb/pull/1440. Is this going to fail if the `shared/` create isn't present?

**Review by philippemnoel — COMMENTED** [review]
very excited

File: src/lib.rs (hunk 2 above)
Comment (philippemnoel) [reply src/lib.rs:49]: Nvm I opened a separate issue for this

**Pull request closed without merging**